### PR TITLE
boards/intel_adsp_cavs25: add rev2 for 2 DSP core variant

### PR DESCRIPTION
**Pull request by kv2019i**
Define revision 2 of the board that has only two DSP cores.

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

**Comment by kv2019i** [comment]
The PR #40861 changed the SoC level default to 4 cores. This covered an issue on SOF side where our existing system to override the Kconfig value for variants with less cores, did not work. Looking at this again, we need something on Zephyr side to support the variant. I considered making a separate board variant, but that seems a bit heavyweight as the HW is otherwise identical. Then I found the board revision feature and used that.

**Review by andyross — COMMENTED** [review]
Is this actually different hardware?  If so, this is obviously the right mechanism.

But note that it's just a kconfig, and you can set that at the app level too.  You can e.g. build slightly different SOF apps differently for different targets, or use -DCONF_FILE=<whatever> at the west command line and it will use that file instead of prj.conf (the default).

Copying new board directories is a little heavyweight aesthetically.  But again if this is actually how the DSP was instantiated we should do it like this.

**Review by andyross — APPROVED** [review]
Sorry, forgot vote.  Those comments are nitpicks and shouldn't block merge.

**Comment by kv2019i** [comment]
@andyross wrote:
> Is this actually different hardware? If so, this is obviously the right mechanism.

Yes, this is a configuration of hardware. E.g. laptops will typically have the 4 core variant, while a desktop chipset (e.g. Z690) will have the 2 core one.

> But note that it's just a kconfig, and you can set that at the app level too. You can e.g. build slightly different SOF apps differently for different targets, or use -DCONF_FILE= at the west command line and it will use that file instead of prj.conf (the default).

I thought about this as well, but given this is property of the hardware, I thought this fits better on Zephyr side. I.e. you should be able to build any Zephyr application for both variants of cavs25.

> Copying new board directories is a little heavyweight aesthetically. But again if this is actually how the DSP was instantiated we should do it like this.

Given only parameter changing is the core count, the "revision" method seemed more appropriate than a completely new board file.

**Comment by marc-hb** [comment]
https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html

> In short:
> - Use devicetree to describe hardware and its boot-time configuration. Examples include peripherals on a board, boot-time clock frequencies, interrupt lines, etc.
> - Use Kconfig to configure software support to build into the final image. Examples include whether to add networking support, which drivers are needed by the application, etc.
>
> In other words, devicetree mainly deals with hardware, and Kconfig with software.

File: boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_2.conf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# board revision 2 = only 2 DSP cores available
+CONFIG_MP_NUM_CPUS=2
```
Comment (marc-hb) [inline boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_2.conf:2]: The missing new line could cause issues when concatenating or other?

File: boards/xtensa/intel_adsp_cavs25/revision.cmake [hunk 2]
```diff
@@ -0,0 +1 @@
+# intentionally empty
```
Comment (marc-hb) [inline boards/xtensa/intel_adsp_cavs25/revision.cmake:1]: I find this very cryptic and the documentation does not seem to mention the possibility of leaving `revision.cmake` empty
https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#custom-revision-cmake-files

Now I plead guilty of ignorance on these topics, but there should be IMHO at least enough comments to give a very vague idea of how this works + a couple specific keywords to starting searching the Zephyr documentation.

**Review by marc-hb — COMMENTED** [review]
Using a board revision and KConfig to match the number of cores  does not seem very intuitive to me but it's not like it's adding technical debt either and I'm too ignorant to suggest anything else so why not. A couple more comments and keywords would not hurt though.

EDIT: the discussion below is _not_ "Outdated" yet. Github is just confused by force pushes as usual.

**Comment by andyross** [comment]
FWIW: MP_NUM_CPUS sits right at the boundary of hardware and software configuration.  In fact, some tests do in fact set this value to control MP behavior (e.g. so you can test "SMP" code paths on uniprocessor situations, or to work around test code that isn't MP-safe).

Ideally we'd have a clean separation between "number of physical cores exposed by the hardware" (in dts) and "which cores should zephyr track at runtime" (in kconfig).  It's possible to imagine asymmetric apps (e.g. the esp32, which has a separate core for the wifi stack) where these are different numbers.

In this case though, it's hardware.  So this is the right choice.

**Comment by andyross** [comment]
To clarify: "In this case it's hardware.  So [putting the change in the board layer which defines the hardware defaults] is the right choice".  A better choice still would be to augment the SMP layer to separate the two notions, but that's a bit out of scope.

File: boards/xtensa/intel_adsp_cavs25/revision.cmake (hunk 2 above)
Comment (kv2019i) [reply boards/xtensa/intel_adsp_cavs25/revision.cmake:1]: There's a build warning if this file does not exist, but we don't have any meaningful content to put here. Let me extend the comment a bit. And open to even better ideas.

File: boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_2.conf (hunk 1 above)
Comment (kv2019i) [reply boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_2.conf:2]: Let me fix that.

File: boards/xtensa/intel_adsp_cavs25/revision.cmake (hunk 2 above)
Comment (marc-hb) [reply boards/xtensa/intel_adsp_cavs25/revision.cmake]: I (finally) read the documentation I'm pretty sure you're supposed to have this in this file:
```
board_check_revision(FORMAT NUMBER)
```
This simply tells the Zephyr build system what revision format you're using, in other words it tells CMake that you are NOT using `intel_adsp_cavs25_A.conf`, `intel_adsp_cavs25_B.conf`, etc.

I'm not sure why it seems to work with an empty file but I would not do that as it's not documented. It could very likely break in the future.

The documentation does not mention the possibility of a "revision-less" board either. It looks like either everything has a revision, or there is a _single_, revision-less board. To be on the safe side and avoid potentially spending hours debugging obscure CMake failures (are CMake issues not obscure ever?), I recommend using a revision always. Then you can go a small extra step and have a nice, consistent and symmetric user interface:
```
west build --board intel_adsp_cavs25@2 # build for 2 cores
west build --board intel_adsp_cavs25@4 # build for 4 cores
```

**Comment by marc-hb** [comment]
The documentation says:
> The optional devicetree overlay `plank_<revision>.overlay` will be added to the common `plank.dts` devicetree file

I see a list of 4 CPUs defined in `zephyr/dts/xtensa/intel/intel_cavs25.dtsi`

So would a new file `boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_2.overlay` file that overrides `intel_cavs25.dtsi` with 2 cores make sense?

@mbolivar-nordic  can you help maybe? Pardon my DT ignorance.

File: boards/xtensa/intel_adsp_cavs25/revision.cmake (hunk 2 above)
Comment (kv2019i) [reply boards/xtensa/intel_adsp_cavs25/revision.cmake]: I tried adding a simple board_check_revision() but all variants I tried broke current SOF way of not specifying a revision. Let me try again. There's an option to specify default revision.

**Comment by MaureenHelm** [comment]
> > Is this actually different hardware? If so, this is obviously the right mechanism.
> 
> Yes, this is a configuration of hardware. E.g. laptops will typically have the 4 core variant, while a desktop chipset (e.g. Z690) will have the 2 core one.
> 
> > But note that it's just a kconfig, and you can set that at the app level too. You can e.g. build slightly different SOF apps differently for different targets, or use -DCONF_FILE= at the west command line and it will use that file instead of prj.conf (the default).
> 
> I thought about this as well, but given this is property of the hardware, I thought this fits better on Zephyr side. I.e. you should be able to build any Zephyr application for both variants of cavs25.
> 
> > Copying new board directories is a little heavyweight aesthetically. But again if this is actually how the DSP was instantiated we should do it like this.
> 
> Given only parameter changing is the core count, the "revision" method seemed more appropriate than a completely new board file.

This should be handled in the SoC-level devicetree, similar to how other SoC families handle part number variants that add, modify, or remove peripheral instances, or have differently sized on-chip memories. You can create a common dtsi that contains all the shared nodes (i.e., two cores), then another dtsi that includes the common one and adds the other two cores. See `dts/arm/nxp/nxp_rt*.dtsi` for example.

**Comment by kv2019i** [comment]
@MaureenHelm wrote:
> This should be handled in the SoC-level devicetree, similar to how other SoC families handle part number variants that add, modify, or remove peripheral instances, or have differently sized on-chip memories. You can create a common dtsi that contains all the shared nodes (i.e., two cores), then another dtsi that includes the common one and adds the other two cores. See `dts/arm/nxp/nxp_rt*.dtsi` for example.

Thanks for feedback. I think we need to take a timeout and spend more time to do this properly. To fix the immediate breakage in SOF CI for all TGL-H systems, I submitted a revert -> https://github.com/zephyrproject-rtos/zephyr/pull/41013 
I.e. we'll limit all cAVS2.5 variants to two cores for now, until a better system is in place. Closing this one.
